lib/storage/src/content_manager/consensus/consensus_wal.rs

```diff
@@ -1,203 +1,288 @@
-                //
-                // `raft::RawNode` would request ridiculous range (e.g., 1 mil entries) with `max_size == Some(0)`,
-                // which, IMO, does not make much sense. 😕
-                //
-                // The current implementation follows `raft::storage::Storage::entries` spec,
-                // but further research/experimentation might be needed.
-                //
-                // See:
-                // - https://docs.rs/raft/latest/raft/storage/trait.Storage.html#tymethod.entries
-                // - https://github.com/tikv/raft-rs/blob/v0.7.0/src/util.rs#L56-L71
-                let max_size = match max_size {
-                    Some(max_size) if max_size < u64::MAX => max_size,
-
-                    // It might also be reasonable to add a hard limit on the total size of returned entries here.
-                    //
-                    // For the same reason, that if `raft::RawNode` request some ginormous range without reasonable
-                    // `max_size` the current `entries` implementation will block consensus thread until the whole
-                    // range is collected.
-                    _ => return true,
-                };
-
-                let first_entry = size == 0;
-                size += u64::from(entry.compute_size());
-
-                size <= max_size || first_entry
-            })
-            .try_collect()?;
+        let offset = self.index_offset()?;
+
+        // Map requested Raft indices to WAL indices
+        let from_wal_index = offset.raft_to_wal(from_raft_index);
+        let until_wal_index = offset.raft_to_wal(until_raft_index);
+
+        // Bound mapped WAL indices between first/last WAL indices
+        let from_wal_index = cmp::max(from_wal_index, offset.wal_index);
+        let until_wal_index = cmp::min(until_wal_index, offset.wal_index + self.wal.num_entries());
+
+        // `Some(u64::MAX)` and `None` are treated as "no max size"
+        // `Some(0)` is treated as "return single entry"
+        //
+        // See:
+        // - https://docs.rs/raft/latest/raft/storage/trait.Storage.html#tymethod.entries
+        // - https://github.com/tikv/raft-rs/blob/v0.7.0/src/util.rs#L56-L71
+        let max_size_bytes = match max_size_bytes {
+            Some(u64::MAX) | None => None,
+            Some(max_size_bytes) => Some(max_size_bytes),
+        };
+
+        let mut size_bytes = 0;
+        let mut entries = Vec::with_capacity(until_wal_index.saturating_sub(from_wal_index) as _);
+
+        for wal_index in from_wal_index..until_wal_index {
+            let entry = self.entry_by_wal_index(wal_index)?;
+
+            size_bytes += u64::from(entry.compute_size());
+            entries.push(entry);
+
+            if let Some(max_size_bytes) = max_size_bytes {
+                if size_bytes >= max_size_bytes {
+                    break;
+                }
+            }
```

Comment:
This likely goes over the limit in many cases.

Shouldn't we try to stay below the limit unless we don't have any items yet?

Maybe we can move the if-statement before `entries.push()`: `if !entries.is_empty() && size_bytes >= max_size_bytes`

Reply:
Hmmm. Maybe. But I've implemented it like this intentionally, so that we always include at least one entry if `max_size_bytes` is 0 (or some other small size, that does not fit a single entry).

Reply:
I understand, but wouldn't the above mentioned alternative achieve the same?

Reply:
Almost, except if you repeatedly request entries with a very small `max_size_bytes` it might never return you anything. Though, I'll verify how original impl handled this and tweak it accordingly.

Reply:
> Almost, except if you repeatedly request entries with a very small max_size_bytes it might never return you anything.

How? Doesn't the `!entries.is_empty()` take care of that?

Reply:
Hmmmm. And you once again right, I think. Ok then. 🤦‍♀️😂

Reply:
Fixed.